### PR TITLE
fix(ignore): ensure normalization of patterns and paths match (fixes #9597)

### LINKED ISSUE #9597: Filenames with extended characters not ignored correctly on macOS
### What happened?

With `.stignore` content:

```
bäd
bad
```
The file `bad` is ignored by `bäd` is not.

I also found a workaround. If you use .stignore` content:

```
ba*d
```
it matches and ignores both files. Possibly `ba?d` would also match, but I have tested it.

I believe this is because the NFD-encoded paths (on macOS) are being matched against NFC-encoded ignore patterns.

The solution is to convert to NFC-encoding before matching.

I have confirmed the bug and a proposed fix by modifying the existing unit test.

Note that some non-ASCII characters work, I presume those that have identical NFC and NFD form.

PR forthcoming.

### Syncthing version

v1.27.8

### Platform & operating system

macOS (64-bit ARM)

### Browser version

_No response_

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by imsodin**
In ignores, normalize the input when parsing it.
When scanning, normalize earlier such that the path is already normalized when checking ignores. This requires splitting normalization of the string from normalization of the file, as we don't want to attempt the latter if the file is ignored.

Closes #9598

### CODE REVIEW

**Comment by imsodin** [comment]
I tried to keep it close to the existing code, to end up with a simple diff. Given the criticality of this code, I'd rather not get it wrong.

File: lib/ignore/ignore.go [hunk 1]
```diff
@@ -387,6 +387,10 @@ func loadParseIncludeFile(filesystem fs.Filesystem, file string, cd ChangeDetect
 }
 
 func parseLine(line string) ([]Pattern, error) {
+	// We use native normalization internally, thus the patterns must match
+	// that to avoid false negative matches.
+	line = osutil.NativeFilename(line)
```
Comment (calmh) [inline lib/ignore/ignore.go:392]: Off the top of my head, does this convert `//comments` to `\\not comments` on Windows?

**Review by calmh — COMMENTED** [review]
Will review fully with a fresher brain

File: lib/ignore/ignore.go (hunk 1 above)
Comment (imsodin) [reply lib/ignore/ignore.go:392]: Oh yes. I actually almost did the same mistake in the scanner, but then realized that was wrong. For some reason that realisation didn't propagate here, probably also need a fresher brain. Will fix tomorrow. And maybe have a look at all places where we do normalisation, if there's an opportunity for something like `osutil.NativeFilename`, just for unicode normalization only.

Comment (imsodin) [reply lib/ignore/ignore.go]: I gave up on the osutil idea: We aren't really consistent in how we apply unicode normalization: In the same contexts we sometimes just leave paths alone, assuming they are already normalized, while in others we apply normalization again. That's a larger project to make that more consistent. I don't see anything that's wrong/buggy, just confusing (and potentially inefficient, though I doubt with any significant impact).

**Comment by imsodin** [comment]
Windows test failure is spurious/unrelated, but macos isn't - looking into that.

**Comment by imsodin** [comment]
Fixed the test input. And removed the newly unused `skip` parameter that deepsource pointed out :)

**Review by calmh — APPROVED** [review]
LGTM!

**Review by calmh — APPROVED** [review]

**Comment by calmh** [comment]
Aight, policy-bot corner case for me to consider at some point,
- I set the thing to require branches to be up to date before merging, because that's that's nice
- I pressed the button to update the branche to enable the merge
- I'm now the author of the merges and the bot disqualifies me from approving the PR :p

**Comment by calmh** [comment]
(Fixed in 30af77f96)

**Pull request merged**